docs-client/src/containers/MethodPage/GrahpqlRequestBody.tsx

```diff
@@ -0,0 +1,190 @@
+import Button from '@material-ui/core/Button';
+import loadable from '@loadable/component';
+
+import {
+  buildClientSchema,
+  GraphQLSchema,
+  getIntrospectionQuery,
+} from 'graphql';
+import TextField from '@material-ui/core/TextField';
+import jsonPrettify from '../../lib/json-prettify';
+import { docServiceDebug } from '../../lib/header-provider';
+
+const graphqlPlaceHolder = `{
+  field(arg: "value") {
+    subField
+  }
+}`;
+const jsonPlaceHolder = jsonPrettify('{"foo":"bar"}');
+
+const QueryEditor = loadable(() => import('./QueryEditor'));
+
+interface Props {
+  requestBodyOpen: boolean;
+  requestBody: string;
+  onEditRequestBodyClick: React.Dispatch<unknown>;
+  onDebugFormChange: (value: string) => void;
+  schemaUrlPath: string;
+}
+
+const toggle = (prev: boolean, override: unknown) => {
+  if (typeof override === 'boolean') {
+    return override;
+  }
+  return !prev;
+};
+
+const parseJson = (requestBody: string) => {
+  let parsedJson;
+  try {
+    parsedJson = JSON.parse(requestBody);
+  } catch (e) {
+    // ignored
+  }
+  return parsedJson;
+};
+
+const GraphqlRequestBody: React.FunctionComponent<Props> = ({
+  requestBodyOpen,
+  requestBody,
+  onEditRequestBodyClick,
+  onDebugFormChange,
+  schemaUrlPath,
+}) => {
+  const [queryOpen, toggleQueryOpen] = useReducer(toggle, true);
+  const [variablesOpen, toggleVariablesOpen] = useReducer(toggle, false);
+
+  const [query, setQuery] = useState('');
+  const [variables, setVariables] = useState('');
+  const [schema, setSchema] = useState<GraphQLSchema | undefined>();
+
+  useEffect(() => {
+    (async () => {
+      const headers: any = {
+        'Content-Type': 'application/json',
+        Accept: 'application/json',
+      };
+      if (process.env.WEBPACK_DEV === 'true') {
+        headers[docServiceDebug] = 'true';
+      }
+      const httpResponse = await fetch(schemaUrlPath, {
+        method: 'POST',
+        headers,
+
+        body: JSON.stringify({
+          operationName: 'IntrospectionQuery',
+          // See https://github.com/graphql/graphiql/blob/8ac05f8b141b6f5cb4449c62ad67a34115490ac8/packages/graphiql/src/utility/introspectionQueries.ts#L16...L22
+          query: getIntrospectionQuery().replace(
+            'subscriptionType { name }',
+            '',
+          ),
+        }),
+      });
+      const result = await httpResponse.json();
+      if (typeof result !== 'string' && 'data' in result) {
+        setSchema(buildClientSchema(result.data));
+      }
+    })();
+  }, [schemaUrlPath]);
+
+  useEffect(() => {
+    const parsedJson = parseJson(requestBody);
+    if (parsedJson === undefined) {
+      return;
+    }
+
+    setQuery(parsedJson.query);
+    if (typeof parsedJson.variables === 'object') {
+      setVariables(jsonPrettify(JSON.stringify(parsedJson.variables)));
+    } else {
+      setVariables(parsedJson.variables);
+    }
+  }, [requestBody]);
+
+  useEffect(() => {
+    const parsedVariables = parseJson(variables);
```

Comment:
Q) Is it possible that `variables` is undefined?

i.e. If `variables` is never modified and only `query` is modified, then is it possible that `query` is never updated?

Reply:
> Q) Is it possible that variables is undefined?

Since the effect is called when variables are changed, it may have an imperfect json string form while typing.

> i.e. If variables is never modified and only query is modified, then is it possible that query is never updated?

Yes... I think so 😂